compiler/next/include/chpl/resolution/scope-types.h

```diff
@@ -0,0 +1,220 @@
+      return false;
+
+    if (moreIds.get()==nullptr && other.moreIds.get()==nullptr)
+      return true;
+
+    // otherwise check the vector elements
+    return *moreIds.get() == *other.moreIds.get();
+  }
+  bool operator!=(const OwnedIdsWithName& other) const {
+    return !(*this == other);
+  }
+};
+struct BorrowedIdsWithName {
+  // TODO: consider storing a variant of ID here
+  // with symbolPath, postOrderId, and tag
+  ID id;
+  const std::vector<ID>* moreIds = nullptr;
+  BorrowedIdsWithName() { }
+  BorrowedIdsWithName(ID id) : id(id) { }
+  BorrowedIdsWithName(const OwnedIdsWithName& o)
+    : id(o.id), moreIds(o.moreIds.get())
+  { }
+  bool operator==(const BorrowedIdsWithName& other) const {
+    return id == other.id &&
+           moreIds == other.moreIds;
+  }
+  bool operator!=(const BorrowedIdsWithName& other) const {
+    return !(*this == other);
+  }
+};
+
+// DeclMap: key - string name,  value - vector of ID of a NamedDecl
+// Using an ID here prevents needing to recompute the Scope
+// if (say) something in the body of a Function changed
+using DeclMap = std::unordered_map<UniqueString, OwnedIdsWithName>;
+
+/**
+  A scope roughly corresponds to a `{ }` block. Anywhere a new symbol could be
+  defined / is defined is a scope.
+
+  While generic instantiations generate something scope-like, the
+  point-of-instantiation reasoning will need to be handled with a different
+  type.
+ */
+// TODO: adjust Conditional to contain Blocks so we can associate
+// scopes 1:1 with these blocks.
+// TODO: also adjust Function to store body in a separate node from args
+// TODO: also adjust Loops to store body separate from index variables
+struct Scope {
+  const Scope* parentScope = nullptr;
+  uast::asttags::ASTTag tag = uast::asttags::NUM_AST_TAGS;
+  bool containsUseImport = false;
+  ID id;
+  DeclMap declared;
+
+  Scope() { }
+
+  bool operator==(const Scope& other) const {
+    bool match =  parentScope == other.parentScope &&
+                  tag == other.tag &&
+                  containsUseImport == other.containsUseImport &&
+                  id == other.id &&
+                  declared.size() == other.declared.size();
+    if (match) {
+      // check also the contents of the maps
+      for (const auto& pair : declared) {
+        UniqueString key = pair.first;
+        const OwnedIdsWithName& val = pair.second;
+        // look up the same key in other
+        auto search = other.declared.find(key);
+        if (search == other.declared.end()) {
+          return false;
+        }
+        // check that they values are the same
+        const OwnedIdsWithName& otherVal = search->second;
+        if (val != otherVal) {
+          return false;
+        }
+      }
+    }
+    return match;
+  }
+  bool operator!=(const Scope& other) const {
+    return !(*this == other);
+  }
+};
+
+// This class supports both use and import
+struct ImportedSymbols {
```

Comment:
How much would you dislike to rename this to `UsedImportedSymbols` or something like that? I really wish we had a generic name for `use`/`import` statements. Reading the code is a bit too challenging with `ImportedSymbols` covering both `use` and `import`

Reply:
Somewhat more confusing: I think we use `UseClause` to represent both `use`/`import`, but here we use `ImportedSymbols` (and other import-based language in names).

Reply:
I've been calling them visibility statements, and I think David L is using that in the renamed version of UseClause

Reply:
Part of it is that David's PR hasn't been merged yet

Reply:
Yes, I'm not really sure what name would be better. `UsedImportedSymbols` reads as "the imported symbols that were used" which isn't what we want. PR #18044 renames UseClause to VisibilityClause but I don't think naming this VisibilitySymbols would be an improvement (since it is too vague IMO) but maybe that is actually better?

Reply:
Maybe `VisibleSymbols`?

Reply:
I'm looking at renaming it to VisibleSymbols and making some related name changes (ResolvedImportScope -> ResolvedVisibilityScope).